[Fix-17177]  Fix cos resource can not list

gaoyan1998 opened the pull request:
## Purpose of the pull request

close #17177 

## Brief change log

fix cos can not list
fix cos list lost folders

gaoyan1998 committed:
Fix cos resource can not list

gaoyan1998 committed:
formate code

ruanwenjun commented:
Please follow the code style of `S3StorageOperator`, it seems this is caused by the directory doesn't end with '/'?

gaoyan1998 commented:
> Please follow the code style of `S3StorageOperator`, it seems this is caused by the directory doesn't end with '/'?

@ruanwenjun hi
I'm sorry I couldn't see where the problem was. Looking at the error, it was a Python E2E test problem, but from the logs, it seems that it was not caused by my pr code. It seems that the installation of the package in testcontainer failed? I don't know if I understand it right

```
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - #5 ERROR: process "/bin/sh -c apt update     && apt install -y software-properties-common     && add-apt-repository ppa:deadsnakes/ppa     && apt update     && apt-get install -y python3.8 libpython3.8-dev python3.8-dev python3.8-distutils     && rm -rf /var/lib/apt/lists/*" did not complete successfully: exit code: 100
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - ------
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] -  > [dolphinscheduler 2/2] RUN apt update     && apt install -y software-properties-common     && add-apt-repository ppa:deadsnakes/ppa     && apt update     && apt-get install -y python3.8 libpython3.8-dev python3.8-dev python3.8-distutils     && rm -rf /var/lib/apt/lists/*:
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80 E: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/noble-updates/main/dep11/Components-amd64.yml.gz  File has unexpected size (216476 != 215904). Mirror sync in progress? [IP: 91.189.91.83 80]
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80    Hashes of expected file:
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80     - Filesize:215904 [weak]
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80     - SHA256:2237a806c02faf36f588e754007941fd5504ac77446236b11f6b3497d6c4bcbd
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80     - SHA1:eb33fce73db3b526008cefb35e92fd854bfec4d3 [weak]
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80     - MD5Sum:4c687407a448453219209b3be60411a2 [weak]
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80    Release file created at: Tue, 13 May 2025 12:44:45 +0000
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80 E: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/noble-updates/universe/dep11/Components-amd64.yml.gz  
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80 E: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/noble-updates/restricted/dep11/Components-amd64.yml.gz  
2025-05-13 14:11:01,088 tc.docker 35 [Thread-3] INFO  [] - 61.80 E: Some index files failed to download. They have been ignored, or old ones used instead.
```

ruanwenjun commented:
> Please follow the code style of `S3StorageOperator`, it seems this is caused by the directory doesn't end with '/'?

If the `resourceAbsolutePath` is an object, can this work?

gaoyan1998 commented:
> > Please follow the code style of `S3StorageOperator`, it seems this is caused by the directory doesn't end with '/'?
> 
> If the `resourceAbsolutePath` is an object, can this work?

The function `listStorageEntity` receive a `String` type Won't pass in the `Object ` type,  Under what circumstances will it become an Object, and what data is this Object?

ruanwenjun commented:
> > > Please follow the code style of `S3StorageOperator`, it seems this is caused by the directory doesn't end with '/'?
> > 
> > 
> > If the `resourceAbsolutePath` is an object, can this work?
> 
> The function `listStorageEntity` receive a `String` type Won't pass in the `Object ` type, Under what circumstances will it become an Object, and what data is this Object?

Sorry, I mean if the path is a file not directory, can this work?

gaoyan1998 commented:
> > > > Please follow the code style of `S3StorageOperator`, it seems this is caused by the directory doesn't end with '/'?
> > > 
> > > 
> > > If the `resourceAbsolutePath` is an object, can this work?
> > 
> > 
> > The function `listStorageEntity` receive a `String` type Won't pass in the `Object ` type, Under what circumstances will it become an Object, and what data is this Object?
> 
> Sorry, I mean if the path is a file not directory, can this work?

not work, But this method is  to list the details of the directory, when will pass a file path?

ruanwenjun commented:
> > > > > Please follow the code style of `S3StorageOperator`, it seems this is caused by the directory doesn't end with '/'?
> > > > 
> > > > 
> > > > If the `resourceAbsolutePath` is an object, can this work?
> > > 
> > > 
> > > The function `listStorageEntity` receive a `String` type Won't pass in the `Object ` type, Under what circumstances will it become an Object, and what data is this Object?
> > 
> > 
> > Sorry, I mean if the path is a file not directory, can this work?
> 
> not work, But this method is to list the details of the directory, when will pass a file path?

You need to follow the interface definition

https://github.com/apache/dolphinscheduler/blob/bdec8b14f6bfda7c4ca21d5449957c2da941b243/dolphinscheduler-storage-plugin/dolphinscheduler-storage-api/src/main/java/org/apache/dolphinscheduler/plugin/storage/api/StorageOperator.java#L132-L140

I have point this before https://github.com/apache/dolphinscheduler/pull/17178#issuecomment-2878463605

gaoyan1998 committed:
fix(storage): fix file and directory query logic in COS storage plugin**

- Optimize the listStorageEntity method to handle queries for files and directories
- Add a check for whether the path ends with `File.separator` to distinguish between file and directory queries
- Add object existence checks to improve query efficiency and accuracy

gaoyan1998 committed:
fix(storage): fix file and directory query logic in COS storage plugin**

- Optimize the listStorageEntity method to handle queries for files and directories
- Add a check for whether the path ends with `File.separator` to distinguish between file and directory queries
- Add object existence checks to improve query efficiency and accuracy

gaoyan1998 committed:
formate code

gaoyan1998 commented:
@ruanwenjun  PTAL 
I've revised the code logic as follows:

Directory Intent: If the input path ends with File.separator, it is directly treated as a directory for querying.
Ambiguous Paths: If the path does not end with File.separator, the system first checks if it matches an existing file. If found, the file is returned; otherwise, it falls back to querying as a directory.

If the path is a file not directory, it`s still work well

Known Issues:

Naming Conflict: If a directory contains both a file and a folder with the same name (and the input path lacks a trailing File.separator), the file will be returned by default. This is intentional—users should append File.separator to explicitly target directories.

Extension Check: The current method to distinguish files from directories (checking for a . in the path) is not entirely reliable. This is a separate issue and will be addressed in a future PR if needed.

SbloodyS committed:
Merge branch 'dev' into fix-cos-bug-17177

ruanwenjun reviewed (approved):
LGTM.

Right now, we can only judge whether the path is file or directory by file name, this is incorrect when the file doesn't have extention, since we don't store the file metadata, we need to provide an additional parameter `isDirectory` in the controller, then listStorageEntity can only used to visit the directory.

SbloodyS reviewed (approved):
+1
